DESIGN.md

```diff
@@ -0,0 +1,164 @@
+
+## Installation, bootstraping
+
+Requires these actions from the user (this should be kept as minimal as possible):
+* Add `NUPM_HOME/bin` to PATH (install location for binary projects)
+* Add `NUPM_HOME/lib` to NU_LIB_DIRS
+* Add `NUPM_HOME/overlays` to NU_LIB_DIRS
+* Make the `nupm` command available somehow (e.g., `use` inside `config.nu`)
+
+WIP: I have another idea in mind, need to think about it. The disadvantage of this is that the default install location is not an overlay.
+We could make `nupm` itself an overlay that adds itself as a command.
+
+There are several approaches:
+* bootstrap using shell script sourced from web (like rustup)
+* embedded inside Nushell's binary
+  * The advantage of this is that it does not require user's config. The PATH and NU_LIB_DIRS could be pre-configured in Nushel
+* (in the future maybe) as a compiled binary (using something like https://github.com/jntrnr/nu_app)
+  * This would allow us to reverse the installation steps: Instead of Nushell installing nupm, we could let user only install nupm which would in turn install Nushell
+
+## Dependency handling
+
+In compiled programming languages, there are two kinds of dependencies: static and dynamic. Static are included statically and compiled when compiling the project,
+dynamic are pre-compiled libraries linked to the project.
+Note that Nushell is [similar to compiled languages][Nushell compiled] rather than typical dynamic languages like Python, so these concepts are relevan for Nushell.
+
+Static dependencies:
+* Advantages: reproducible, does not rely on system files (no more missing random.so.2), higher performance (allows joint optimization of dependencies and project itself)
+* Disadvatage: increased compile time, binary size, can easily end up with multiple versions of the same library (hello Nushell dependencies)
+
+Dynamic dependencies are the opposite basically. Note that Nushell currently supports only static dependencies, but we might be able to add the "linking" feature at some point.
+
+We might want `nupm`support both types of dependencies.
+
+## Package repository 
+
+Packages need to be stored somewhere. There should be one central "official" location (see https://github.com/NixOS/nixpkgs for inspiration).
+
+Additionally, user should be able to add 3rd party repositories as well as install local and other packages (e.g., from the web, just pointing at URL),
+as long as it has `METADATA_FILE` telling `nupm` what to do.
+
+## API / CLI Interface
```

Comment:
It kinda depends on metadata which is to be done buuut...

If we agree on having both short and long package descriptions, all commands displaying some output (`list`, `install`, `search`, `uninstall`, `update`) should by default display the short description and flag (one of them):

- `--long (-l)`
- `--long-desc (-l)`
- `--long-description (-l)`

To display the long description instead of the short one

And I think that this comment concludes the flags discussion for now 😄

Reply:
what about 345da642362f2e0d259f349c8b1cbe673e12f91f? :wink:

Reply:
I would replace "give" with "print" but generally speaking, fine 😃